### ISSUE TITLE
WCW Nitro: the mysterious wrong textures

### ISSUE THREAD

**Issue by olivieryuyu** [gonetz/GLideN64#179]
![hardware](https://cloud.githubusercontent.com/assets/8987490/5358766/f6ce08ba-7fb7-11e4-80dc-e24b5b74975d.jpg)

HLE and LLE has the same issue. Only software plugins (HLE/LLE) have not this issue.

It must be a framebuffer/TMEM issue then (CI framebuffer index?)

**Comment by olivieryuyu** [comment]
![hhh](https://cloud.githubusercontent.com/assets/8987490/8914931/2e6d08c6-34a2-11e5-8b86-7dfcd99e31e9.jpg)

render framebuffer as texture improves the situation, without solving it

**Comment by oddMLan** [comment]
Native res x1 helps but textures are still messed up heavily
![gliden64_nitro64_001](https://cloud.githubusercontent.com/assets/7278372/15261952/379892b0-1914-11e6-902d-135c52512194.png)

**Comment by olivieryuyu** [comment]
second screen is now only garbage, regression!

**Comment by olivieryuyu** [comment]
Btw only jabo software solve the issue, both in lle and hle

**Comment by olivieryuyu** [comment]
why software plugin solve the case in both LLE and HLE?

**Comment by gonetz** [comment]
Would be nice to get sources and find why. May be it uses a hack, like Glide64 does.

**Comment by olivieryuyu** [comment]
a hack only on software HLE but not hardware HLE? that would be odd. You can switch Jabo mode from hardware to software. The correct image appears and then goes black.

still a mystery :)

**Comment by olivieryuyu** [comment]
actually you are right, it works simply with copy fb to rdram with jabo. in HLE hardware as well. 

Damn!

**Comment by olivieryuyu** [comment]
reading some docs, I have now an idea... :) 

Are the letters 4 bits textures please?

**Comment by gonetz** [comment]
Yes, text uses 4bit CI textures.

**Comment by olivieryuyu** [comment]
It must be a loading texture issue. Z64 LLE doesn't work with this game. As well the ucode is standard.

I believed for a while it was a kind of framebuffer issue (as Yoshi). But libreto LLE implementation doesn't have the issue and it is based on angrylion plugin. It is hardware with the same capacity in terms of framebuffer than GlideN64.

So it must be a RDP issue. My feeling is that it relates to 4 bits textures. Source code of sprite shows that they are dealt separately than other textures. For instance there is some difference as: 

_Textureis4Bit: NumWordsPerLine = (SubImageWidth * 4  + ((SubImageWidth & 0x01) ? 4 : 0) + (((SubImageWidth + SourceImageOffsetS) & 0x01) ? 4 : 0) + 63) >> 6;_

and for other textures:

_TextureisNot4Bit: NumWordsPerLine = (SubImageWidth  * (TileBitSize[SourceImageBitSize] == 32 ? 16 : TileBitSize[SourceImageBitSize])  + 63) >> 6;)_

By the way I read in the programming manual that:

4-bit textutre should be loaded as 16-bit types and then rendered as 4-bit types (is it what is refered as SubImageWidth * 4 by the way?).

And also:

 For color index (CI) textures, the texture is stored in the lower half of Tmem, and the Texture/Color Look-Up Table (TLUT) is stored in the upper half of Tmem. For 4-bit CI textures, the texels (or indices) addressed in the lower half of Tmem have the 4-bit palette number for the tile prepended to create an 8-bit address into the upper half of Tmem. Since four texels are addressed simultaneously, there must be four (usually identical) TLUTs stored in the upper half of Tmem across the four banks.

For 4-bit CI textures, the palette effectively selects one of sixteen possible tables, each table having sixteen entries. Each table is aligned on 16-word boundaries. Note that there are two choices for the texel type that resides in the TLUT: 16-bit RGBA, or 16-bit IA. The type is selected using the gDPSetTextureLUT() command. This command also configures the Tmem as shown in the following figure. Because of this, CI textures cannot be combined with other texture types in two-cycle mode. 

Not sure I stand correct here but I believe that this is around this that the issue stands.

**Comment by olivieryuyu** [comment]
LLE and HLE has the same issue. It is not a HLE issue, it is a RDP issue. The microcode is the normal one as far as I can see.

**Comment by olivieryuyu** [comment]
would it be possible to check the logs between angrylion and your plugin to see the difference in LLE? That would be useful no?

**Comment by angrylion7** [comment]
The contents of the TMEM match perfectly between GlideN64 and an older version of my plugin, which could handle this scene, so Load* commands in GlideN64 should be good enough, and they also match my older plugin very well code-wise. 
This scene also doesn't use any frame buffer effects. I also excluded many other possibilities during my testing, by tweaking various othermodes bits, by making the two tile indices used non-neighbours in their 0...7 index space, by using just one tile instead of two for all operations (which is possible in this particular display list), by inserting a bogus SetMaskImage command, which is absent in this DL, because it doesn't use z-buffering. 
So I went on and created a testrom which only renders the very first texrect of this scene, it was wrong in GlideN64 too. The only notable thing about this CI-8 texrect is that dsdx = 1.0 (0x400), while dtdy = 2.0 (0x800). So I lowered its dtdy to 1.0 and 0.5 and increased YL accordingly, and the divergence from my plugin was gone! Then I increased dtdy to 3.0 and decreased YL accordingly, and the bug was there but less pronounced. So it should be an issue with the "rendering" stage.
Since the LSB of integer parts and whole fractional parts of T coordinates are also always zero in this scene while dtdy is always 0x800, in my plugin this not only effectively surpasses triangular filtering, it also means that "dword interleave" when fetching texels from the TMEM never ever happens.
So maybe it's just a bug in how GlideN64 renders CI-4 and CI-8 textures when dtdy >= 2.0 (or even simply > 1.0), or it may be somehow connected with this "dword interleave", which is somehow not emulated well for textures of the above mentioned formats and for big dtdy's.

**Comment by gonetz** [comment]
@angrylion7 , thanks for the analysis! I'll look at this problem after solving current regressions.

**Comment by gonetz** [comment]
I found the problem.

Lets consider LLE case.

First texture image:
gDPSetTextureImage( G_IM_FMT_CI, G_IM_SIZ_8b, 704, 0x00121140 );
That is game set 8bit CI texture with origin at address 0x00121140 and width 704.

Then this texture image is loaded to TMEM using LoadTile commands. 
LoadTile "copies" rectangular sub-image from large texture image in RDRAM to TMEM.
Tile size is 64x32

This tile is rendered by texrect with size 64x16. Looks strange: why use tile which has doubled height?
The texrect uses dtdy = 2.0, that is each time Y coordinate increased by 1, T coordinate increased by 2, so 16 pixel height of rectangle mapped to 32 texel height of texture. In other words, half of texture texels skipped.

So where the problem? The problem is that actual image width is not 704, it is 1408
How LoadTile works in this case? For example, the very first LoadTile with uls = 0 and ult = 0.
line 0: skip 704*0 texels from origin and loads 64 texels of texture image to TMEM
line 1: skip 704*1 texels from origin, loads another 64 texels
line 2: skip 704*2 texels from origin, loads another 64 texels
and so on.
Each line with odd index in tile is wrong, because after skipping 704 texels we will be still on the same line, not on the next one as it should be.
No problems for software render: it skips lines with odd indices from the tile because of dtdy = 2.
Large problem for hardware render, because it uses continuous texture coordinates from 0 to 1. and thus half pixels get wrong texels. May be implementation of #1603 will allow to fix it without hacks.

**Comment by olivieryuyu** [comment]
finally the mystery is cracked!!!

May be not solved but at least identified :)

**Comment by olivieryuyu** [comment]
Would it be possible that #114 is the same issue?

**Comment by olivieryuyu** [comment]
@gonetz

Do I understand that you need to be able to use non normalized coordinates for texture rectangle?

If so what not using:

https://www.khronos.org/opengl/wiki/Rectangle_Texture

Coud it be?

**Comment by gonetz** [comment]
Rectangle_Texture is an interesting feature. I did not know about it, thanks!

Use of Rectangle_Texture by itself can't solve the problem. As the description says, "It could certainly be emulated by using the textureSize function to get the size and manually normalize a texture coordinate." That is, we can do the same without Rectangle_Texture. It may help to implement #1603, since non-normalized fetching would be more natural with use of Rectangle_Texture.

In short, this particular issue can be fixed either with a hack (easy way) or with serious modification of texturing code.
